src/main/java/mate/academy/bookshop/service/impl/BookServiceImpl.java

```diff
@@ -79,4 +87,28 @@ public List<BookDto> search(Pageable pageable, BookSearchParameters bookSearchPa
     public boolean existsByIsbn(String isbn) {
         return bookRepository.existsByIsbn(isbn);
     }
+
+    @Override
+    public List<BookDtoWithoutCategoryIds> findAllByCategoryIds(Pageable pageable, Long id) {
+        validateCategory(id);
+        return bookRepository.findAllByCategoryId(pageable, id).stream()
+                .map(bookMapper::toDtoWithoutCategories)
+                .toList();
+    }
+
+    private void validateCategory(Long id) {
+        if (!categoryRepository.existsById(id)) {
+            throw new EntityNotFoundException("Category with such id: "
+                    + id + " not exist");
+        }
+    }
+
+    private Set<Category> getCategories(Set<Long> categoriesIds) {
+        Set<Category> categories = new HashSet<>();
+        categoriesIds.stream()
+                .peek(this::validateCategory)
+                .map(categoryRepository::findById)
+                .forEach(category -> categories.add(category.get()));

```

Comment:
Please investigate if there is more efficient possibility to make this logic, because you need to make N requests to DB in current implementation